### DISCUSSION TITLE
Please provide pointers to setup Selenium Hub or equivalent for Taiko/Gauge with JS

### DISCUSSION THREAD

**Discussion by james-k-mathew** [getgauge/taiko#1872] [General]
We are using Taiko + Gauge with JS
We need to run many scripts in parallel across multiple machines
Is there a way to setup Selenium grid for the (Taiko + Gauge with JS) or
is there any other equivalent way to do it

**Comment by harimadusumilli** [comment]
Yes, any pointers and documentation will be helpful. I am also looking for this documentation and ways to set it up.

**Comment by NivedhaSenthil** [comment]
selenoid can be used for this usecase I feel...there is [plugin](https://github.com/saikrishna321/taiko-selenoid) for the same 

@saikrishna321 might have more context on this

**Comment by saikrishna321** [comment]
@harimadusumilli @james-k-mathew I use [Selenoid](https://aerokube.com/selenoid/latest/) as grid and trigger the test in parallel. Each test will spin a new cluster in parallel.

**Comment by james-k-mathew** [comment]
@NivedhaSenthil @saikrishna321 , thank you very much, will look into it

**Reply by james-k-mathew** [reply]
I have installed selenoid, I took a look at the taiko-selenoid plugin. It is not clear where I should give the selenoid url. Can you please help

@saikrishna321 I also took a look at

https://github.com/getgauge/taiko/issues/1356

Can the code snippet mentioned on 20 Sep 2020 take url which is other than localhost

**Reply by james-k-mathew** [reply]
@saikrishna321 we have a critical requirement to run tests across multiple machines. The current architecture of Selenoid points to run from one machine only (with multiple instances running in docker containers) Please let us know how we can configure the Selenoid url:port to run the Taiko scripts. Also please let us know how this can be done across multiple machines.

**Reply by saikrishna321** [reply]
@jamesgeorge007 Will improve the plugin to accept url and port.

**Reply by saikrishna321** [reply]
Can u give more info on Multiple machines ?

**Reply by james-k-mathew** [reply]
Hi @saikrishna321  we need to run tests across a set of AWS instances. With each instance running multiple tests in parallel. Something similar to how it is done using selenium hub. It is a critical requirement for us and any help towards that direction will be very helpful.

Also, Please let us know if the url and port can be configured in the current version of plugin.

**Reply by saikrishna321** [reply]
Please use the latest version of taiko-selenoid this has already been fixed and released.

**Reply by james-k-mathew** [reply]
Hi @saikrishna321 I installed the latest version over the Selenoid which was installed in one AWS instance. I am able to open two instances of browser and able to view both browser instances using selenoid-ui url. I understand multiple tests can also be run in this way in that AWS instance.
    We have multiple instances of AWS and we want to leverage all of them for running the tests. Can you please suggest how this can be done.

**Reply by nseshadr** [reply]
Is there an update on this issue? We also have a requirement to run on multiple AWS instances in parallel and see how the load is distributed and reports consolidated

**Reply by james-k-mathew** [reply]
Hi @saikrishna321 Is there any update on the extension to multiple instances. It will greatly help in the automatic load distribution and getting a consolidated result. In the current setup each instance is behaving as a separate unit. Each instance will have to be assigned tests initially and post the test run, results will have to be got from each instance and consolidated.

**Reply by saikrishna321** [reply]
Hey! Have u folks looked at ggr which acts as load balancer https://github.com/aerokube/ggr

**Reply by james-k-mathew** [reply]
Hi @saikrishna321   I tried the installation of Ggr in one AWS instance and the slave node of Selenoid in another AWS instance. After installation, when I try
http://<Ggr url>:4444/quota
I am able to see the slave Selenoid node details and the attached browser details.
When I try
http://<Ggr url>:4444/wd/hub
I am getting the "404 page not found" message

1. Is there anything different which should be done for Ggr installation
2. Once the Ggr install is done, please let me know how I can trigger Taiko tests into that.

**Reply by james-k-mathew** [reply]
hi @saikrishna321 whatever params I try to give for the openBrowser() to connect to the Selenoid instance, I get the following error

'Error: Selenoid Session creation Failed connect ECONNREFUSED 127.0.0.1:4444 with data {"desiredCapabilities":{"browserName":"chrome","browserVersion":"86.0","selenoid:options":{"sessionTimeout":"3m","enableVnc":true}}}

It always tries to connect to 127.0.0.1
Please let me know the way to connect to a different IP

**Reply by saikrishna321** [reply]
Hi! I haven't tried ggr before. Looking at the error looks like your selenoid.config.js is missing the port and host.

**Reply by james-k-mathew** [reply]
hi @saikrishna321     I changed to this configuration.   AWS Linux instance with SelenoidTaiko plugin -> AWS Linux instance with Ggr -> AWS Linux instance with slave Selenoid

Now from the AWS instance with SelenoidTaiko plugin, I am able to create instance in slave Selenoid by using the openBrowser command of Taiko. It was also possible to hit the slave Selenoid directly by using its url and port.

Earlier I was having the SelenoidTaiko plugin in a Windows machine and using openBrowser was leading to errors.

**Reply by saikrishna321** [reply]
@james-k-mathew Great you got it working. Can you please add this info to the taiko-selenoid ReadMe?

**Reply by james-k-mathew** [reply]
@saikrishna321 I was using a slightly different configuration. Selenoid, taiko-selenoid plugin, Gauge and Taiko in the same LInux instance. I tried to run many tests in parallel mode this time. After some time the tests were failing with afterSuite failure and screen shot taking completely failing.
    After I tried to see where the problem could be, I realized that the test was timing out at 3 minutes, whether all the tests in all the threads were completed or not.
    Saw that the sessionTimeout variable in the node_modules/taiko-selenoid/lib/index.js file was set to 3m. After I changing it to 30m the tests started running properly without getting timed out/afterSuite failing/not taking screenshot.

**Reply by saikrishna321** [reply]
@james-k-mathew The latest master on taiko-selenoid has this fix to configure the session timeout using _setSelenoidConfig_

https://github.com/saikrishna321/taiko-selenoid/commit/daf00668761089972405791af518795920ef83ef#diff-d40fbd8da63e458ea729a9b8e0ac654aa6bd68e07f014acfc93a6506f1ffe863
